src/compiler/checker.ts

```diff
@@ -6924,20 +6946,13 @@ namespace ts {
             let signatureList: Signature[];
             let types = (<UnionType>type).types;
             for (let current of types) {
-                // The signature set of all constituent type with call signatures should match
-                // So number of signatures allowed is either 0 or 1
-                if (signatureList &&
-                    getSignaturesOfStructuredType(current, SignatureKind.Call).length > 1) {
-                    return undefined;
-                }
-
                 let signature = getNonGenericSignature(current);
                 if (signature) {
                     if (!signatureList) {
                         // This signature will contribute to contextual union signature
                         signatureList = [signature];
                     }
-                    else if (!compareSignatures(signatureList[0], signature, /*compareReturnTypes*/ false, compareTypes)) {
+                    else if (!compareSignatures(signatureList[0], signature, /*partialMatch*/ false, /*ignoreReturnTypes*/ true, compareTypes)) {

```

Comment:
Possibly, but it is a bit different here in that were trying to find the most informative signature, not the least common denominator. For example:

```typescript
var f: ((x: string) => void) | (() => void) = x => { ... };
```

Here you'd really like the contextual signature to be `(x: string) => void` so `x` can be contextually typed. But then it get complicated in situations like this:

```typescript
var f: ((x: string) => void) | ((x: number) => void) | () => void) = x => { ... };
```

Here we'd have to go with `() => void` and no contextual type for `x`.

Anyway, not clear that there is a simple solution so I'm leaving the existing rule in place.